cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,869 @@
+    bool nulls_at_start,
+    scalar const* row_delta,
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr) const
+  {
+    CUDF_EXPECTS(!row_delta || cudf::have_same_types(orderby, *row_delta),
+                 "Orderby column and row_delta must have the same type.",
+                 cudf::data_type_error);
+    return window_bounds<OrderbyT>(orderby, grouping, nulls_at_start, row_delta, stream, mr);
+  }
+
+  template <typename OrderbyT,
+            CUDF_ENABLE_IF(cuda::std::is_same_v<OrderbyT, cudf::string_view> &&
+                           (WindowTag == window_tag::CURRENT_ROW ||
+                            WindowTag == window_tag::UNBOUNDED))>
+  [[nodiscard]] std::unique_ptr<column> operator()(
+    column_view const& orderby,
+    std::optional<preprocessed_group_info> const& grouping,
+    bool nulls_at_start,
+    scalar const* row_delta,
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr) const
+  {
+    CUDF_EXPECTS(!row_delta,
+                 "Not expecting window range to have value for string-based window calculation");
+    return window_bounds<OrderbyT>(orderby, grouping, nulls_at_start, row_delta, stream, mr);
+  }
+
+  template <typename OrderbyT, CUDF_ENABLE_IF(!is_supported<OrderbyT>())>
+  std::unique_ptr<column> operator()(column_view const&,
+                                     std::optional<preprocessed_group_info> const&,
+                                     bool,
+                                     scalar const*,
+                                     rmm::cuda_stream_view,
+                                     rmm::device_async_resource_ref) const
+  {
+    CUDF_FAIL("Unsupported rolling window type.", cudf::data_type_error);
+  }
+};
+}  // namespace rolling
+
+namespace {
+template <class... Ts>
+struct match : Ts... {
+  using Ts::operator()...;
+};
+template <class... Ts>
+match(Ts...) -> match<Ts...>;
+}  // namespace
+
+template <rolling::direction Direction>
+[[nodiscard]] inline std::unique_ptr<column> make_range_window(
+  column_view const& orderby,
+  std::optional<rolling::preprocessed_group_info> const& grouping,
+  order order,
+  null_order null_order,
+  range_window_type window,
+  rmm::cuda_stream_view stream,
+  rmm::device_async_resource_ref mr)
+{
+  CUDF_FUNC_RANGE();
+  bool const nulls_at_start = (order == order::ASCENDING && null_order == null_order::BEFORE) ||
+                              (order == order::DESCENDING && null_order == null_order::AFTER);
+
+  using tag_t   = rolling::window_tag;
+  auto dispatch = [&](auto&& clamper, scalar const* row_delta) {
+    return type_dispatcher(
+      orderby.type(), clamper, orderby, grouping, nulls_at_start, row_delta, stream, mr);
+  };
+  return std::visit(
+    match{
+      [&](bounded_closed win) {
+        if (order == order::ASCENDING) {
+          return dispatch(
+            rolling::range_window_clamper<Direction, tag_t::BOUNDED_CLOSED, order::ASCENDING>{},
```

Comment:
Could use `bounded_closed` as the template argument, but arguably it will make the code inside the clamper struct more complicated.